### PR TITLE
[SYCLomatic]Add cugraph helpers - 

### PR DESCRIPTION
**Pull request by abhilash1910**
Add helpers for : 
cudaGraphGetNode, cudaGraphRootNodes, cudaGraphNodeGetDependentNodes, cudaGraphNodeGetDependencies, cudaGraphDebugDotPrint

cc @yihanwg , @zhimingwang36, @tangjj11  and also @daiyaan-ahmed6 .

### CODE REVIEW

**Comment by abhilash1910** [comment]
@daiyaan-ahmed6  @zhiweij1 could you help review ? Thanks

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 1]
```diff
@@ -133,5 +133,75 @@ static void add_dependencies(dpct::experimental::command_graph_ptr graph,
   }
 }
 
+///  Auxiliary function to query graph nodes.
+/// \param [in] graph A pointer to the command graph.
+/// \param [in] get_func Lambda function for overloading
+/// \param [in] expectednumberOfNodes The number of nodes
+/// to be added
+inline std::vector<sycl::ext::oneapi::experimental::node> get_nodes_impl(
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp]: It seems to be an internal function; how about putting it into `namespace detail`?

**Review by zhiweij1 — COMMENTED** [review]
Will these newly added functions be used to migrate some CUDA APIs?
If so, could you show the mapping?

File: clang/runtime/dpct-rt/include/dpct/graph.hpp (hunk 1 above)
Comment (abhilash1910) [reply clang/runtime/dpct-rt/include/dpct/graph.hpp]: Yes can be added there, Will update.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 2]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp]:
```suggestion
```

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 3]
```diff
@@ -25,6 +26,25 @@ typedef sycl::ext::oneapi::experimental::command_graph<
 typedef sycl::ext::oneapi::experimental::node *node_ptr;
 
 namespace detail {
+
+///  Auxiliary function to query graph nodes.
+/// \param [in] graph A pointer to the command graph.
+/// \param [in] get_func Lambda function for overloading
+/// \param [in] expectednumberOfNodes The number of nodes
+/// to be added
+inline std::vector<sycl::ext::oneapi::experimental::node> get_nodes_impl(
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp:34]: Please use clang-format to format the new added code.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 4]
```diff
@@ -25,6 +26,25 @@ typedef sycl::ext::oneapi::experimental::command_graph<
...
+    std::function<std::vector<sycl::ext::oneapi::experimental::node>()> get_func,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp]: Can we simplify the parameter?
```suggestion
   const std::vector<sycl::ext::oneapi::experimental::node> &nodes,
```

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 5]
```diff
@@ -133,5 +153,66 @@ static void add_dependencies(dpct::experimental::command_graph_ptr graph,
...
+/// \Queries the nodes present in graph 
+/// \param [in] graph A pointer to the command graph.
+/// \param [in] numberOfNodes number of Nodes in graph.
+
+inline std::vector<sycl::ext::oneapi::experimental::node> get_nodes(
+    const dpct::experimental::command_graph_ptr graph,
+    std::size_t numberOfNodes) {
+
+    return detail::get_nodes_impl(std::function<std::vector<sycl::ext::oneapi::experimental::node>()>(
+        [&]() { return graph->get_nodes(); }), numberOfNodes);
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp]: How about putting `detail::get_nodes_impl` in `dpct::experimental` namespace and migrate the code to
```suggestion
    get_nodes_impl(graph->get_nodes(), numberOfNodes)
```
directly?
Same suggestion for below 3 helper functions.

File: clang/runtime/dpct-rt/include/dpct/graph.hpp [hunk 6]
```diff
@@ -133,5 +153,66 @@ static void add_dependencies(dpct::experimental::command_graph_ptr graph,
...
+        [&]() { return node->get_predecessors(); }), numberOfPredecessorNodes);
+    
+}
+
+/// \Prints the graph 
+/// \param [in] graph A pointer to the command graph.
+/// \param [in] path FilePath Path to the output file.
+/// \param [in] verbose If true, print additional information about the nodes such
+/// as kernel args or memory access where applicable.
+static inline void print_graph(const dpct::experimental::command_graph_ptr graph,
```
Comment (zhiweij1) [inline clang/runtime/dpct-rt/include/dpct/graph.hpp:209]: Do we have other reason for adding this helper function?
Can we migrate the code to `graph->print_graph(path, verbose)` directly?

**Comment by abhilash1910** [comment]
@daiyaan-ahmed6 please help to pick this up  as tests are needed, thanks

**Comment by zhimingwang36** [comment]
Close this PR,  as no test case available and not clear how it is used for.

**Pull request closed without merging**